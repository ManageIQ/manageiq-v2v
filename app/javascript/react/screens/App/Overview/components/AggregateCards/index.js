import MigrationsNotStarted from './MigrationsNotStarted/MigrationsNotStarted';
import MigrationsInProgress from './MigrationsInProgress/MigrationsInProgress';
import MigrationsComplete from './MigrationsComplete/MigrationsComplete';
import InfrastructureMappings from './InfrastructureMappings/InfrastructureMappings';

export {
  MigrationsNotStarted,
  MigrationsInProgress,
  MigrationsComplete,
  InfrastructureMappings
};
