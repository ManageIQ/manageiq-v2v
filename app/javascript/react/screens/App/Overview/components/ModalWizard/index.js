import ModalWizard from './ModalWizard';
import ModalWizardBody from './ModalWizardBody';

ModalWizard.Body = ModalWizardBody;

export default ModalWizard;
