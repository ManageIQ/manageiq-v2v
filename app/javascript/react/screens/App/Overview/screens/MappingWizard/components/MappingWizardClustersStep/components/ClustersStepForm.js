import React from 'react';
import PropTypes from 'prop-types';
import { bindMethods } from 'patternfly-react';

import DualPaneMapper from '../../DualPaneMapper/DualPaneMapper';
import DualPaneMapperList from '../../DualPaneMapper/DualPaneMapperList';
import DualPaneMapperCount from '../../DualPaneMapper/DualPaneMapperCount';
import DualPaneMapperListItem from '../../DualPaneMapper/DualPaneMapperListItem';
import ClustersStepTreeView from './ClustersStepTreeView';

import { sourceClustersFilter } from '../MappingWizardClustersStepSelectors';

class ClustersStepForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedTargetCluster: null,
      selectedSourceClusters: [],
      selectedMapping: null
    };

    bindMethods(this, [
      'selectSourceCluster',
      'selectTargetCluster',
      'addMapping',
      'selectMapping',
      'removeMapping',
      'removeAll'
    ]);
  }

  selectSourceCluster(sourceCluster) {
    this.setState(prevState => {
      const isAlreadySelected = prevState.selectedSourceClusters.some(
        cluster => cluster.id === sourceCluster.id
      );
      if (isAlreadySelected) {
        return {
          selectedSourceClusters: prevState.selectedSourceClusters.filter(
            cluster => cluster.id !== sourceCluster.id
          )
        };
      }
      return {
        selectedSourceClusters: [
          ...prevState.selectedSourceClusters,
          sourceCluster
        ]
      };
    });
  }

  selectTargetCluster(targetCluster) {
    this.setState(() => ({ selectedTargetCluster: targetCluster }));
  }

  addMapping() {
    const { input } = this.props;
    this.setState(prevState => {
      const clustersStepMappings = input.value;

      const mappingExistsForSelectedTargetCluster = clustersStepMappings.some(
        targetClusterWithSourceClusters =>
          targetClusterWithSourceClusters.id ===
          prevState.selectedTargetCluster.id
      );

      if (mappingExistsForSelectedTargetCluster) {
        const updatedMappings = clustersStepMappings.map(
          targetClusterWithSourceClusters => {
            if (
              targetClusterWithSourceClusters.id ===
              prevState.selectedTargetCluster.id
            ) {
              return {
                ...targetClusterWithSourceClusters,
                nodes: targetClusterWithSourceClusters.nodes.concat(
                  prevState.selectedSourceClusters.map(sourceCluster => ({
                    ...sourceCluster,
                    text: sourceCluster.name,
                    icon: 'fa fa-file-o'
                  }))
                )
              };
            }
            return targetClusterWithSourceClusters;
          }
        );
        input.onChange(updatedMappings);
      } else {
        const targetClusterWithSourceClusters = {
          ...prevState.selectedTargetCluster,
          text: prevState.selectedTargetCluster.name,
          state: {
            expanded: true
          },
          selectable: true,
          selected: false,
          nodes: prevState.selectedSourceClusters.map(sourceCluster => ({
            ...sourceCluster,
            text: sourceCluster.name,
            icon: 'fa fa-file-o'
          }))
        };
        input.onChange([
          ...clustersStepMappings,
          targetClusterWithSourceClusters
        ]);
      }
      return {
        selectedTargetCluster: null,
        selectedSourceClusters: []
      };
    });
  }

  selectMapping(selectedMapping) {
    const { input } = this.props;
    input.onChange(
      input.value.map(mapping => {
        if (mapping.id === selectedMapping.id) {
          return { ...mapping, selected: !mapping.selected };
        } else if (mapping.id !== selectedMapping.id && mapping.selected) {
          return { ...mapping, selected: false };
        }
        return mapping;
      })
    );
    this.setState(() => ({ selectedMapping }));
  }

  removeMapping() {
    const { input } = this.props;
    this.setState(prevState => {
      const clustersStepMappings = input.value;
      input.onChange(
        clustersStepMappings.filter(
          targetClusterWithSourceClusters =>
            !(
              targetClusterWithSourceClusters.id ===
              prevState.selectedMapping.id
            )
        )
      );
      return {
        selectedMapping: null
      };
    });
  }

  removeAll() {
    const { input } = this.props;
    input.onChange([]);
  }

  render() {
    const {
      sourceClusters,
      targetClusters,
      isFetchingSourceClusters,
      isFetchingTargetClusters,
      input
    } = this.props;

    const {
      selectedTargetCluster,
      selectedSourceClusters,
      selectedMapping
    } = this.state;

    return (
      <div className="dual-pane-mapper-form">
        <DualPaneMapper
          handleButtonClick={this.addMapping}
          validMapping={
            !(
              selectedTargetCluster &&
              (selectedSourceClusters && selectedSourceClusters.length > 0)
            )
          }
        >
          {sourceClusters && (
            <DualPaneMapperList
              listTitle="Source Clusters"
              loading={isFetchingSourceClusters}
            >
              {sourceClustersFilter(sourceClusters, input.value).map(item => (
                <DualPaneMapperListItem
                  item={item}
                  key={item.id}
                  selected={
                    selectedSourceClusters &&
                    selectedSourceClusters.some(
                      sourceCluster => sourceCluster.id === item.id
                    )
                  }
                  handleClick={this.selectSourceCluster}
                  handleKeyPress={this.selectSourceCluster}
                />
              ))}
              <DualPaneMapperCount
                selectedItems={selectedSourceClusters.length}
                totalItems={
                  sourceClustersFilter(sourceClusters, input.value).length
                }
              />
            </DualPaneMapperList>
          )}
          {targetClusters && (
            <DualPaneMapperList
              listTitle="Target Clusters"
              loading={isFetchingTargetClusters}
            >
              {targetClusters.map(item => (
                <DualPaneMapperListItem
                  item={item}
                  key={item.id}
                  selected={
                    selectedTargetCluster &&
                    selectedTargetCluster.id === item.id
                  }
                  handleClick={this.selectTargetCluster}
                  handleKeyPress={this.selectTargetCluster}
                />
              ))}
            </DualPaneMapperList>
          )}
        </DualPaneMapper>
        {input.value && (
          <ClustersStepTreeView
            mappings={input.value}
            selectMapping={this.selectMapping}
            removeMapping={this.removeMapping}
            removeAll={this.removeAll}
            selectedMapping={selectedMapping}
          />
        )}
      </div>
    );
  }
}

ClustersStepForm.propTypes = {
  input: PropTypes.object,
  sourceClusters: PropTypes.arrayOf(PropTypes.object),
  targetClusters: PropTypes.arrayOf(PropTypes.object),
  isFetchingSourceClusters: PropTypes.bool,
  isFetchingTargetClusters: PropTypes.bool
};

export default ClustersStepForm;
