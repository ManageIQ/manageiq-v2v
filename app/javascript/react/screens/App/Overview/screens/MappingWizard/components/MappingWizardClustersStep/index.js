import { connect } from 'react-redux';
import MappingWizardClustersStep from './MappingWizardClustersStep';
import * as MappingWizardClustersStepActions from './MappingWizardClustersStepActions';

import reducer from './MappingWizardClustersStepReducer';

export const reducers = { mappingWizardClustersStep: reducer };

const mapStateToProps = ({ mappingWizardClustersStep }, ownProps) => ({
  ...mappingWizardClustersStep,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(
  mapStateToProps,
  MappingWizardClustersStepActions,
  mergeProps
)(MappingWizardClustersStep);
