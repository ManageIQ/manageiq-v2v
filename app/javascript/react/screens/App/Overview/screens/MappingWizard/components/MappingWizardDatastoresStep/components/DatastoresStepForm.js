import React from 'react';
import PropTypes from 'prop-types';
import { bindMethods } from 'patternfly-react';
import cx from 'classnames';

import DualPaneMapper from '../../DualPaneMapper/DualPaneMapper';
import DualPaneMapperList from '../../DualPaneMapper/DualPaneMapperList';
import DualPaneMapperCount from '../../DualPaneMapper/DualPaneMapperCount';
import DualPaneMapperListItem from '../../DualPaneMapper/DualPaneMapperListItem';
import MappingWizardTreeView from '../../MappingWizardTreeView/MappingWizardTreeView';

import { sourceDatastoreFilter } from '../MappingWizardDatastoresStepSelectors';
import {
  diskSpaceInfo,
  sourceDatastoreInfo,
  targetDatastoreInfo
} from './helpers';

class DatastoresStepForm extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      selectedSourceDatastores: [],
      selectedTargetDatastore: null,
      selectedMapping: null
    };

    bindMethods(this, [
      'selectSourceDatastore',
      'selectTargetDatastore',
      'addDatastoreMapping',
      'selectMapping',
      'removeMapping',
      'removeAll'
    ]);
  }

  componentWillReceiveProps(nextProps) {
    if (
      this.props.selectedCluster &&
      nextProps.selectedCluster.id !== this.props.selectedCluster.id
    ) {
      this.setState(() => ({
        selectedSourceDatastores: [],
        selectedTargetDatastore: null
      }));
    }
  }

  selectSourceDatastore(sourceDatastore) {
    this.setState(prevState => {
      const isAlreadySelected = prevState.selectedSourceDatastores.some(
        datastore => datastore.id === sourceDatastore.id
      );
      if (isAlreadySelected) {
        return {
          selectedSourceDatastores: prevState.selectedSourceDatastores.filter(
            datastore => datastore.id !== sourceDatastore.id
          )
        };
      }
      return {
        selectedSourceDatastores: [
          ...prevState.selectedSourceDatastores,
          sourceDatastore
        ]
      };
    });
  }

  selectTargetDatastore(targetDatastore) {
    this.setState(() => ({ selectedTargetDatastore: targetDatastore }));
  }

  addDatastoreMapping() {
    const { input, selectedCluster, selectedClusterMapping } = this.props;

    const { selectedTargetDatastore } = this.state;

    const noMappingForTargetCluster = !input.value.some(
      datastoreMapping => datastoreMapping.id === selectedClusterMapping.id
    );

    const addingToExistingMapping = input.value.some(
      targetClusterDatastoreMappings =>
        targetClusterDatastoreMappings.nodes.some(
          datastoreMapping => datastoreMapping.id === selectedTargetDatastore.id
        )
    );

    this.setState(prevState => {
      if (input.value.length === 0 || noMappingForTargetCluster) {
        input.onChange([
          ...input.value,
          {
            ...selectedClusterMapping,
            text: selectedClusterMapping.name,
            selectable: false,
            nodes: [
              {
                ...prevState.selectedTargetDatastore,
                text: diskSpaceInfo(
                  prevState.selectedTargetDatastore,
                  prevState.selectedSourceDatastores
                ),
                selectable: true,
                selected: false,
                state: {
                  expanded: true
                },
                nodes: prevState.selectedSourceDatastores.map(datastore => ({
                  ...datastore,
                  text: datastore.name,
                  icon: 'fa fa-file-o',
                  sourceClusterId: selectedCluster.id
                }))
              }
            ]
          }
        ]);
      } else {
        input.onChange(
          input.value.map(datastoreMapping => {
            if (datastoreMapping.id !== selectedClusterMapping.id) {
              return datastoreMapping;
            } else if (addingToExistingMapping) {
              return {
                ...datastoreMapping,
                nodes: datastoreMapping.nodes.map(mapping => {
                  if (mapping.id === prevState.selectedTargetDatastore.id) {
                    return {
                      ...mapping,
                      text: diskSpaceInfo(
                        mapping,
                        mapping.nodes.concat(prevState.selectedSourceDatastores)
                      ),
                      nodes: mapping.nodes.concat(
                        prevState.selectedSourceDatastores.map(datastore => ({
                          ...datastore,
                          text: datastore.name,
                          icon: 'fa fa-file-o',
                          sourceClusterId: selectedCluster.id
                        }))
                      )
                    };
                  }
                  return mapping;
                })
              };
            }
            return {
              ...datastoreMapping,
              nodes: datastoreMapping.nodes.concat({
                ...prevState.selectedTargetDatastore,
                text: diskSpaceInfo(
                  prevState.selectedTargetDatastore,
                  prevState.selectedSourceDatastores
                ),
                selectable: true,
                selected: false,
                state: {
                  expanded: true
                },
                nodes: prevState.selectedSourceDatastores.map(datastore => ({
                  ...datastore,
                  text: datastore.name,
                  icon: 'fa fa-file-o',
                  sourceClusterId: selectedCluster.id
                }))
              })
            };
          })
        );
      }
      return {
        selectedTargetDatastore: null,
        selectedSourceDatastores: []
      };
    });
  }

  selectMapping(selectedMapping) {
    const { input } = this.props;

    input.onChange(
      input.value.map(targetClusterDatastoreMappings => {
        const updatedMappings = targetClusterDatastoreMappings.nodes.map(
          datastoreMapping => {
            if (datastoreMapping.id === selectedMapping.id) {
              return {
                ...datastoreMapping,
                selected: !datastoreMapping.selected
              };
            } else if (
              datastoreMapping.id !== selectedMapping.id &&
              datastoreMapping.selected
            ) {
              return { ...datastoreMapping, selected: false };
            }
            return datastoreMapping;
          }
        );
        return { ...targetClusterDatastoreMappings, nodes: updatedMappings };
      })
    );
    this.setState(() => ({ selectedMapping }));
  }

  removeMapping() {
    const { input } = this.props;
    const { selectedMapping } = this.state;
    const datastoresStepMappings = input.value;

    const updatedDatastoresStepMappings = datastoresStepMappings
      .map(targetClusterWithDatastoreMappings => {
        const containsMappingToRemove = targetClusterWithDatastoreMappings.nodes.some(
          mapping => mapping.id === selectedMapping.id
        );
        if (!containsMappingToRemove) {
          return targetClusterWithDatastoreMappings;
        }
        const updatedDatastoreMappings = targetClusterWithDatastoreMappings.nodes.filter(
          datastoreMapping => !(datastoreMapping.id === selectedMapping.id)
        );
        if (updatedDatastoreMappings.length === 0) {
          return undefined;
        }
        return {
          ...targetClusterWithDatastoreMappings,
          nodes: updatedDatastoreMappings
        };
      })
      .filter(item => item !== undefined);

    input.onChange(updatedDatastoresStepMappings);
    this.setState(() => ({ selectedMapping: null }));
  }

  removeAll() {
    const { input } = this.props;
    input.onChange([]);
  }

  render() {
    const {
      sourceDatastores,
      targetDatastores,
      isFetchingSourceDatastores,
      isFetchingTargetDatastores,
      input,
      selectedCluster
    } = this.props;
    const {
      selectedSourceDatastores,
      selectedTargetDatastore,
      selectedMapping
    } = this.state;

    const classes = cx('dual-pane-mapper-form', {
      'is-hidden': !selectedCluster
    });

    return (
      <div className={classes}>
        <DualPaneMapper
          handleButtonClick={this.addDatastoreMapping}
          validMapping={
            !(
              selectedTargetDatastore &&
              (selectedSourceDatastores && selectedSourceDatastores.length > 0)
            )
          }
        >
          <DualPaneMapperList
            listTitle="Source Datastores"
            loading={isFetchingSourceDatastores}
          >
            {sourceDatastores &&
              sourceDatastoreFilter(sourceDatastores, input.value).map(item => (
                <DualPaneMapperListItem
                  item={item}
                  text={sourceDatastoreInfo(item)}
                  key={item.id}
                  selected={
                    selectedSourceDatastores &&
                    selectedSourceDatastores.some(
                      sourceDatastore => sourceDatastore.id === item.id
                    )
                  }
                  handleClick={this.selectSourceDatastore}
                  handleKeyPress={this.selectSourceDatastore}
                />
              ))}
            <DualPaneMapperCount
              selectedItems={selectedSourceDatastores.length}
              totalItems={
                sourceDatastoreFilter(sourceDatastores, input.value).length
              }
            />
          </DualPaneMapperList>
          <DualPaneMapperList
            listTitle="Target Datastores"
            loading={isFetchingTargetDatastores}
          >
            {targetDatastores &&
              targetDatastores.map(item => (
                <DualPaneMapperListItem
                  item={item}
                  text={targetDatastoreInfo(item, input.value)}
                  key={item.id}
                  selected={
                    selectedTargetDatastore &&
                    selectedTargetDatastore.id === item.id
                  }
                  handleClick={this.selectTargetDatastore}
                  handleKeyPress={this.selectTargetDatastore}
                />
              ))}
          </DualPaneMapperList>
        </DualPaneMapper>
        <MappingWizardTreeView
          mappings={input.value}
          selectMapping={this.selectMapping}
          removeMapping={this.removeMapping}
          removeAll={this.removeAll}
          selectedMapping={selectedMapping}
        />
      </div>
    );
  }
}

export default DatastoresStepForm;

DatastoresStepForm.propTypes = {
  input: PropTypes.object,
  selectedCluster: PropTypes.object,
  selectedClusterMapping: PropTypes.object,
  sourceDatastores: PropTypes.array,
  targetDatastores: PropTypes.array,
  isFetchingSourceDatastores: PropTypes.bool,
  isFetchingTargetDatastores: PropTypes.bool
};
