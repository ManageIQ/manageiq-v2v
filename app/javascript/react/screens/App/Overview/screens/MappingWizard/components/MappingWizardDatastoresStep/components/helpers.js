import numeral from 'numeral';

export const datastoreUsedSpace = (datastore = {}) =>
  datastore.total_space - datastore.free_space;

export const totalUsedSpace = (datastores = []) =>
  datastores.reduce(
    (totalSpace, datastore) => (totalSpace += datastoreUsedSpace(datastore)),
    0
  );

export const targetDatastoreAvailableSpace = (
  targetDatastore,
  datastoresStepMappings
) => {
  const datastoresMappings = datastoresStepMappings.reduce(
    (mappings, targetClusterWithDatastoresMappings) =>
      mappings.concat(targetClusterWithDatastoresMappings.nodes),
    []
  );

  const matchingDatastoresMapping = datastoresMappings.find(
    targetDatastoreWithSourceDatastores =>
      targetDatastoreWithSourceDatastores.id === targetDatastore.id
  );

  if (datastoresMappings.length > 0 && matchingDatastoresMapping) {
    return (
      targetDatastore.free_space -
      totalUsedSpace(matchingDatastoresMapping.nodes)
    );
  }
  return targetDatastore.free_space;
};

export const sourceDatastoreInfo = sourceDatastore =>
  sprintf(
    __('%s (%s)'),
    sourceDatastore.name,
    numeral(datastoreUsedSpace(sourceDatastore)).format('0.00b')
  );

export const targetDatastoreInfo = (targetDatastore, datastoresStepMappings) =>
  sprintf(
    __('%s (%s avail)'),
    targetDatastore.name,
    numeral(
      targetDatastoreAvailableSpace(targetDatastore, datastoresStepMappings)
    ).format('0.00b')
  );

export const targetDatastoreTreeViewInfo = (
  targetDatastore,
  sourceDatastores
) => {
  const { total_space, free_space } = targetDatastore;
  const availableSpace = free_space - totalUsedSpace(sourceDatastores);

  return sprintf(
    __('%s (%s total, %s avail)'),
    targetDatastore.name,
    numeral(total_space).format('0.00b'),
    numeral(availableSpace).format('0.00b')
  );
};

export const errorMessage = __(
  'The size of the selected source datastores exceeds the available space in the target datastore'
);
