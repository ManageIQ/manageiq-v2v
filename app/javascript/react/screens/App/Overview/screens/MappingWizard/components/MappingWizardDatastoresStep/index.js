import { connect } from 'react-redux';
import MappingWizardDatastoresStep from './MappingWizardDatastoresStep';
import * as MappingWizardDatastoresStepActions from './MappingWizardDatastoresStepActions';

import reducer from './MappingWizardDatastoresStepReducer';

export const reducers = { mappingWizardDatastoresStep: reducer };

const mapStateToProps = ({ mappingWizardDatastoresStep }, ownProps) => ({
  ...mappingWizardDatastoresStep,
  ...ownProps.data
});

const mergeProps = (stateProps, dispatchProps, ownProps) =>
  Object.assign(stateProps, ownProps.data, dispatchProps);

export default connect(
  mapStateToProps,
  MappingWizardDatastoresStepActions,
  mergeProps
)(MappingWizardDatastoresStep);
