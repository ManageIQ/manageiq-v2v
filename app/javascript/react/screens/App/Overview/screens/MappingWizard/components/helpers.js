import { V2V_TARGET_PROVIDERS } from '../MappingWizardConstants';

export const getClusterOptions = clusterMappings => {
  const sourceClustersWithAssociatedTargetClusters = clusterMappings.reduce(
    (mappings, targetClusterWithSourceClusters) => {
      const { nodes: sourceClusters, ...targetCluster } = targetClusterWithSourceClusters;
      const sourceToTargetMappings = sourceClusters.map(sourceCluster => ({
        sourceCluster,
        targetCluster,
        sourceClusterMappedToTargetCluster: {
          name: `${sourceCluster.name} (${targetCluster.name})`,
          id: sourceCluster.id
        }
      }));
      return mappings.concat(sourceToTargetMappings);
    },
    []
  );

  return sourceClustersWithAssociatedTargetClusters.map(
    ({ sourceClusterMappedToTargetCluster }) => sourceClusterMappedToTargetCluster
  );
};

export const multiProviderTargetLabel = (providerType, wizardStep) => {
  switch (providerType) {
    case V2V_TARGET_PROVIDERS[1].id:
      switch (wizardStep) {
        case 'cluster':
          return __('Target Provider \\ Project');
        case 'storage':
          return __('Target Provider \\ Volume Type');
        case 'network':
          return __('Target Project \\ Network');
        default:
          return null;
      }
    default:
      switch (wizardStep) {
        case 'cluster':
          return __('Target Provider \\ Datacenter \\ Cluster');
        case 'storage':
          return __('Target Datastores');
        case 'network':
          return __('Target Network');
        default:
          return null;
      }
  }
};

export const updateMappings = (mappings, clusterStepMappings) =>
  mappings.filter(mapping => {
    const potentialAssociatedClusterStepMappings = clusterStepMappings.filter(
      clusterStepMapping => clusterStepMapping.id === mapping.id
    );

    if (potentialAssociatedClusterStepMappings.length === 0) {
      return false;
    }

    const associatedClusterMapping = potentialAssociatedClusterStepMappings.find(clusterStepMapping => {
      const { nodes: sourceClusters } = clusterStepMapping;
      const sourceClusterIds = sourceClusters.map(sourceCluster => sourceCluster.id);
      const { nodes: mappedSourceStoragesOrNetworks } = mapping.nodes[0];

      return mappedSourceStoragesOrNetworks.every(storageOrNetwork =>
        sourceClusterIds.some(id => id === storageOrNetwork.sourceClusterId)
      );
    });

    return associatedClusterMapping;
  });
