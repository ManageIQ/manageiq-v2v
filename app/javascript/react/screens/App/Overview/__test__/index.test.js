import React from 'react';
import thunk from 'redux-thunk';
import promiseMiddleware from 'redux-promise-middleware';
import { createStore, applyMiddleware, combineReducers } from 'redux';
import { mount } from 'enzyme';
import { Provider } from 'react-redux';
import { initialState } from '../overview.fixtures';
import OverviewContainer, { reducers } from '../index';

import { coreComponents } from '../../../../../components';
import componentRegistry from '../../../../../components/componentRegistry';

jest.mock('../../../../../components/componentRegistry');
componentRegistry.registerMultiple(coreComponents);

describe('Overview integration test', () => {
  const middlewares = [thunk, promiseMiddleware()];
  const generateStore = () =>
    createStore(
      combineReducers({ ...reducers }),
      {
        overview: initialState
      },
      applyMiddleware(...middlewares)
    );

  const mountComponent = store =>
    mount(
      <Provider store={store}>
        <OverviewContainer />
      </Provider>
    );

  it('should mount the Overview with mapStateToProps reduced', () => {
    const store = generateStore();
    const wrapper = mountComponent(store);

    // * query the unconnected component to assert reduced props
    // * because we are async loading Overview with react-loadable, we need to
    //   refer to it as 'LoadableComponent'
    const component = wrapper.find('LoadableComponent');

    expect(component.props()).toMatchSnapshot();
  });
});
