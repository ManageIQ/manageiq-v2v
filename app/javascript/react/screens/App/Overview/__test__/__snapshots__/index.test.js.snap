// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Overview integration test should mount the Overview with mapStateToProps reduced 1`] = `
Object {
  "continueToPlanAction": [Function],
  "fetchTransformationMappingsAction": [Function],
  "fetchTransformationMappingsUrl": "/api/dummyMappings",
  "fetchTransformationPlanRequestsAction": [Function],
  "fetchTransformationPlanRequestsUrl": "/api/dummyPlanRequests",
  "fetchTransformationPlansAction": [Function],
  "fetchTransformationPlansUrl": "/api/dummyTransformationPlans",
  "hideMappingWizard": false,
  "hidePlanWizard": false,
  "isFetchingTransformationMappings": true,
  "isFetchingTransformationPlanRequests": true,
  "isFetchingTransformationPlans": true,
  "isRejectedTransformationMappings": false,
  "mappingWizardVisible": false,
  "planWizardVisible": false,
  "showMappingWizardAction": [Function],
  "showPlanWizardAction": [Function],
  "transformationMappings": Array [],
}
`;
